source/extensions/filters/network/common/redis/codec_impl.cc

```diff
@@ -391,6 +397,148 @@ void DecoderImpl::parseSlice(const Buffer::RawSlice& slice) {
       break;
     }
 
+    case State::InlineStart: {
+      ENVOY_LOG(trace, "parse slice: InlineStart: {}", buffer[0]);
+      if (buffer[0] == '\r') {
+        state_ = State::LF;
+      } else if (std::isspace(buffer[0])) {
+        // Discard whitespace
+      } else {
+        RespValuePtr pending_value = std::make_unique<RespValue>();
+        pending_value->type(RespType::BulkString);
+
+        if (buffer[0] == '"') {
+          state_ = State::InlineStringQuoted;
+        } else if (buffer[0] == '\'') {
+          state_ = State::InlineStringSingleQuoted;
+        } else {
+          pending_value->asString().push_back(buffer[0]);
+          state_ = State::InlineString;
+        }
+
+        size_t n = pending_value_stack_.front().value_->asArray().size();
+        pending_value_stack_.front().value_->asArray().push_back(*pending_value);
+        pending_value_stack_.push_front({&pending_value_stack_.front().value_->asArray()[n], n});
+      }
+
+      remaining--;
+      buffer++;
+      break;
+    }
+
+    case State::InlineDelimiter: {
+      ENVOY_LOG(trace, "parse slice: InlineDelimiter: {}", buffer[0]);
+      if (buffer[0] == '\r') {
+        state_ = State::LF;
+      } else if (std::isspace(buffer[0])) {
+        state_ = State::InlineStart;
+      } else {
+        throw ProtocolError("unbalanced quotes in request");
+      }
+
+      remaining--;
+      buffer++;
+      break;
+    }
+
+    case State::InlineString: {
+      ENVOY_LOG(trace, "parse slice: InlineString: {}", buffer[0]);
+
+      if (buffer[0] == '\r') {
+        pending_value_stack_.pop_front();
+        state_ = State::LF;
+      } else if (std::isspace(buffer[0])) {
+        pending_value_stack_.pop_front();
+        state_ = State::InlineStart;
+      } else if (buffer[0] == '"') {
+        throw ProtocolError("unbalanced quotes in request");
+      } else {
+        pending_value_stack_.front().value_->asString().push_back(buffer[0]);
+      }
+
+      remaining--;
+      buffer++;
+      break;
+    }
+
+    case State::InlineStringQuoted: {
+      ENVOY_LOG(trace, "parse slice: InlineStringQuoted: {}", buffer[0]);
+
+      if (buffer[0] == '\r') {
+        throw ProtocolError("unbalanced quotes in request");
+      } else if (buffer[0] == '"') {
+        pending_value_stack_.pop_front();
+        state_ = State::InlineDelimiter;
+      } else {
+        if (buffer[0] == '\\' && remaining > 1) {
+          char c;
+          switch (buffer[1]) {
+          case '\\':
+            c = '\\';
+            break;
+          case 'x':
+            if (remaining > 3 && std::isxdigit(buffer[2]) && std::isxdigit(buffer[3])) {
```

Comment:
There may be a problem with looking at the later bytes in the buffer because the request data can be divided into multiple slices.
https://github.com/envoyproxy/envoy/blob/1810bff57495ba5a8862a7d86a8e818d1c68fa78/source/extensions/filters/network/common/redis/codec_impl.cc#L333

So it's possible that the 'x' character is the last byte of one slice, and the later bytes come in the next slice.

Wondering whether it would be possible to add more states in the state machine without looking at future bytes.

Reply:
Good catch! I'll add additional states instead.

Reply:
Done.